In [4]:
import pandas
from sklearn import linear_model

In [10]:
df = pandas.read_csv("cars.csv")
df.head()


,Car,Model,Volume,Weight,CO2
0,Toyoty,Aygo,1000,790,99
1,Mitsubishi,Space Star,1200,1160,95
2,Skoda,Citigo,1000,929,95
3,Fiat,500,900,865,90
4,Mini,Cooper,1500,1140,105


In [7]:
X = df[['Weight', 'Volume']]
y = df['CO2']

regr = linear_model.LinearRegression()
regr.fit(X, y)

#predict the CO2 emission of a car where the weight is 2300kg, and the volume is 1300cm3:
predictedCO2 = regr.predict([[2300, 1300]])

print(predictedCO2)

[107.2087328]


/opt/anaconda3/lib/python3.13/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


# Coefficient

The coefficient is a factor that describes the relationship with an unknown variable.
In this case, we can ask for the coefficient value of weight against CO2, and for volume against CO2. The answer(s) we get tells us what would happen if we increase, or decrease, one of the independent values.

In [10]:
print(regr.coef_)

[0.00755095 0.00780526]


The result array represents the coefficient values of weight and volume.

Weight: 0.00755095
Volume: 0.00780526

These values tell us that if the weight increase by 1kg, the CO2 emission increases by 0.00755095g.

And if the engine size (Volume) increases by 1 cm3, the CO2 emission increases by 0.00780526 g.

I think that is a fair guess, but let test it!

We have already predicted that if a car with a 1300cm3 engine weighs 2300kg, the CO2 emission will be approximately 107g.

What if we increase the weight with 1000kg (from 2300 to 3300) what will be the CO2 emission?

Ans: 107.2087328 + (1000 * 0.00755095) = 114.75968

In [11]:
predictedCO2 = regr.predict([[3300, 1300]])

print(predictedCO2)

[114.75968007]


/opt/anaconda3/lib/python3.13/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [13]:
#Checking how many observations are in the current file
len(df)

36

## Experimenting
To experiment further with this dataset, I've created an expanded version of it using ChatGPT, feeding it the original dataset and asking it to expand it to 100 observations.
This will allow me to divide the data into train and test, make manipulations in the data, and learn more about multiple linear regression

The function here is similar to the one in the simple linear regression, but instead of one X we have 2 X's, representing two predictors.
Running the code below, we learn that the intercept = 84.215:

$\hat{y} = 84.215 + 0.00805X + 0.00429X$

In [16]:
df = pandas.read_csv('cars_100_observations.csv')

In [18]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42,shuffle=True)

In [25]:
#Pasted the code again here to play with different values of Weight and Volume

import pandas as pd

regr = linear_model.LinearRegression()
regr.fit(X_train, y_train)

y_pred_test = regr.predict(X_test) #generate predictions for test
comparison_test = pd.DataFrame({
    'Actual': y_test,
    'Predicted': y_pred_test
})

print(comparison_test)
print("Coefficients:", regr.coef_)
print("Intercept:", regr.intercept_)

    Actual   Predicted
35     120  106.162301
13      94  101.152581
26     104  104.783275
30     115  106.137684
16      99  102.062149
31     117  106.843891
21      99  104.284220
12      99   97.453233
Coefficients: [0.00804928 0.00428741]
Intercept: 84.21502006695282


Next, we compare the performance of the model across the train and test datasets.
We do this with two metrics:
- R² - tells us how much variance the model explains
- MAE - tells us how much off, on average, the predictions are.

**The results we got are:**
Train R²: 0.2595386654370383
Test R²: 0.32941109624012743
Train MAE: 4.647253536064064
Test MAE: 6.335639624274503

**Learnings:**
- The model has relatively low R² on both datasets, explaining approximately 26% of the variance in the training data and 33% in the test data.
- Interestingly, test R² is slightly higher than train R². This can happen, particularly with a small test set, because of variation in which observations are selected.
- MAE is higher on the test set (6.34) than the training set (4.65), meaning that predictions for unseen observations are further from the actual CO₂ values on average.
- Overall, the relatively low R² suggests that Volume and Weight alone provide limited explanatory power for CO₂, and there may be room to improve the model by including additional relevant predictors.

In [26]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

print("Train R²:", r2_score(y_train, y_pred_train))
print("Test R²:", r2_score(y_test, y_pred_test))

print("Train MAE:", mean_absolute_error(y_train, y_pred_train))
print("Test MAE:", mean_absolute_error(y_test, y_pred_test))

Train R²: 0.2595386654370383
Test R²: 0.32941109624012743
Train MAE: 4.647253536064064
Test MAE: 6.335639624274503
